src/backend/gl/src/command.rs

```diff
@@ -480,84 +480,121 @@ impl CommandBuffer {
+                    | ChannelType::Snorm
+                    | ChannelType::Ufloat
+                    | ChannelType::Sfloat
+                    | ChannelType::Srgb
+                    | ChannelType::Uscaled
+                    | ChannelType::Sscaled => clear_cmds
+                        .push(Command::ClearBufferColorF(index, unsafe {
+                            clear.value.color.float32
+                        })),
+                    ChannelType::Uint => clear_cmds
+                        .push(Command::ClearBufferColorU(index, unsafe {
+                            clear.value.color.uint32
+                        })),
+                    ChannelType::Sint => clear_cmds
+                        .push(Command::ClearBufferColorI(index, unsafe {
+                            clear.value.color.sint32
+                        })),
+                };
+
+                if let Some(mask) = color_mask {
+                    clear_cmds.push(Command::SetColorMask(Some(index), mask));
                 }
-                None
-            })
-            .collect::<Vec<_>>();
+            } else {
+                // Clear depth-stencil target
+                let depth = if view_format.is_depth()
+                    && attachment.ops.load == pass::AttachmentLoadOp::Clear
+                {
+                    Some(unsafe { clear.value.depth_stencil.depth })
+                } else {
+                    None
+                };
+
+                // Only reset depth mask if it was non writable
+                let depth_mask = self.cache.depth_mask.filter(|mask| !mask);
+
+                let stencil = if view_format.is_stencil()
+                    && attachment.stencil_ops.load == pass::AttachmentLoadOp::Clear
+                {
+                    Some(unsafe { clear.value.depth_stencil.stencil })
+                } else {
+                    None
+                };
+
+                let stencil_mask = self
+                    .cache
+                    .stencil_mask
+                    .filter(|mask| mask.front != !0 || mask.back != !0);
+
+                // Temporarily reset masks as they may prevent buffer clear in gl
+                if depth_mask.is_some() {
+                    clear_cmds.push(Command::SetDepthMask(true));
+                }
+                if stencil_mask.is_some() {
+                    clear_cmds.push(Command::SetStencilMask(!0));
+                }
+
+                if depth.is_some() || stencil.is_some() {
+                    clear_cmds.push(Command::ClearBufferDepthStencil(depth, stencil));
+                }
+
+                // Restore masks if they were reset
+                if let Some(mask) = depth_mask {
+                    clear_cmds.push(Command::SetDepthMask(mask));
+                }
+                if let Some(mask) = stencil_mask {
+                    clear_cmds.push(Command::SetStencilMaskSeparate(mask));
+                }
+            }
+        }
```

Comment:
can we reorder the code in such a way that we could just live-add the clear commands (with `self.push_cmd`) instead of collecting them and the flushing them? I.e. can we move the `BindFrameBuffer` and `DrawBuffers` before this giant block?

Reply:
I was about to do that, but you can't call `self.push_cmd` as it requires `&mut self` and self was already borrowed to access `self.pass_cache`.

Reply:
do you need to borrow the cache though?